### PR TITLE
Adds COLORED_COMPILER_OUTPUT CMAKE flag

### PR DESCRIPTION
**Pull request by dtasev**
**Description of work.**
Adds `-DCOLORED_OUTPUT` true/false CMake flag.

Code taken from: https://medium.com/@alasher/colored-c-compiler-output-with-ninja-clang-gcc-10bfe7f2b949

**To test:**
Compile something with an error locally. Output should have colours.

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by DTasev** [comment]
@martyngigg I wasn't sure where to put the code, is `buildconfig/CMake/GNUSetup.cmake` the right file for something like this?

**Comment by martyngigg** [comment]
@DTasev That is the correct place for gcc/clang-specific options.

File: buildconfig/CMake/GNUSetup.cmake [hunk 1]
```diff
@@ -115,3 +115,13 @@ if( CMAKE_COMPILER_IS_GNUCXX )
     set( CMAKE_SHARED_LIBRARY_LINK_CXX_FLAGS "${CMAKE_SHARED_LIBRARY_LINK_CXX_FLAGS} -static-libgcc -static-libstdc++" )
   endif()
 endif()
+
+option (COLORED_OUTPUT "Always produce ANSI-colored output (GNU/Clang only)." TRUE)
+
+if (${COLORED_OUTPUT})
```
Comment (martyngigg) [inline buildconfig/CMake/GNUSetup.cmake]: Just a small style point for cmake. You can avoid the `${}` brackets in testing variables as cmake will expand them automatically, for example this is equivalent to `if(COLORED_OUTPUT)`. We tend to use the latter style as it's easier to read.

File: buildconfig/CMake/GNUSetup.cmake [hunk 2]
```diff
@@ -115,3 +115,13 @@ if( CMAKE_COMPILER_IS_GNUCXX )
...
+    if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
```
Comment (martyngigg) [inline buildconfig/CMake/GNUSetup.cmake]: The same can be done for `STREQUAL`: `if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")` is equivalent but is easier to read with the quotes and brackets.

File: buildconfig/CMake/GNUSetup.cmake [hunk 3]
```diff
@@ -115,3 +115,13 @@ if( CMAKE_COMPILER_IS_GNUCXX )
     set( CMAKE_SHARED_LIBRARY_LINK_CXX_FLAGS "${CMAKE_SHARED_LIBRARY_LINK_CXX_FLAGS} -static-libgcc -static-libstdc++" )
   endif()
 endif()
+
+option (COLORED_OUTPUT "Always produce ANSI-colored output (GNU/Clang only)." TRUE)
```
Comment (martyngigg) [inline buildconfig/CMake/GNUSetup.cmake]: Maybe a name that is slightly more specific like `COLORED_COMPILER_OUTPUT`?

**Review by martyngigg — CHANGES_REQUESTED** [review]
This will be very nice. A few style points to be consistent with our other CMake usage.

File: buildconfig/CMake/GNUSetup.cmake (hunk 2 above)
Comment (DTasev) [reply buildconfig/CMake/GNUSetup.cmake]: Can I use the `CMAKE_COMPILER_IS_GNUCXX` variable instead of doing the STREQUAL for GNU?

Comment (martyngigg) [reply buildconfig/CMake/GNUSetup.cmake]: https://cmake.org/cmake/help/v3.14/variable/CMAKE_COMPILER_IS_GNUCXX.html?highlight=cmake_compiler_is_gnucxx seems to recommend using `CMAKE_CXX_COMPILER_ID`

File: buildconfig/CMake/GNUSetup.cmake [hunk 4]
```diff
@@ -115,3 +115,13 @@ if( CMAKE_COMPILER_IS_GNUCXX )
     set( CMAKE_SHARED_LIBRARY_LINK_CXX_FLAGS "${CMAKE_SHARED_LIBRARY_LINK_CXX_FLAGS} -static-libgcc -static-libstdc++" )
   endif()
 endif()
+
+option (COLORED_COMPILER_OUTPUT "Always produce ANSI-colored output (GNU/Clang only)." TRUE)
+
+if(COLORED_COMPILER_OUTPUT)
+  if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+    add_compile_options(-fdiagnostics-color=always)
```
Comment (martyngigg) [inline buildconfig/CMake/GNUSetup.cmake]: If the `COLORED_COMPILER_OUTPUT` is `False` should we set `-fdiagnostics-color=never`?

Comment (DTasev) [reply buildconfig/CMake/GNUSetup.cmake]: Is an explicit else necessary? I think the flag being false will make it behave as it has until before this PR

Comment (martyngigg) [reply buildconfig/CMake/GNUSetup.cmake]: I was just wondering whether a developer would want `False` to mean definitely turn this off and not auto detect whether it should be on.

Comment (DTasev) [reply buildconfig/CMake/GNUSetup.cmake]: That is true, I'll add the never for `False` as it makes sense

**Review by martyngigg — APPROVED** [review]
Looks good to me.

**Pull request merged**